In [1]:
import pandas as pd

# Объединение таблиц Pandas DataFrame
Предположим, для анализа предоставлены несколько файлов с логами (например, по 1 файлу за каждый день), которые следует анализировать совместно.

In [2]:
Log_1 = pd.read_csv('log_1.csv', sep=';')
Log_2 = pd.read_csv('log_2.csv', sep=';')

In [3]:
Log_1

ID  SOURCE                 DATE
0   720251  yandex  17.10.2019 11:17:17
1   720250  google  17.10.2019 11:09:32
2   720255  yandex  17.10.2019 12:04:34
3   720257  google  17.10.2019 12:17:11
4   720258  yandex  17.10.2019 12:40:07
5   720281  yandex  17.10.2019 14:10:31
6   720290  yandex  17.10.2019 14:37:11
7   720297  yandex  17.10.2019 15:21:57
8   720301  yandex  17.10.2019 16:20:13
9   720303  yandex  17.10.2019 16:26:25
10  720253  google  17.10.2019 11:53:01
11  720256  yandex  17.10.2019 12:12:44
12  720262  yandex  17.10.2019 12:53:55

In [4]:
Log_2

ID  SOURCE                 DATE
0  720474  yandex  18.10.2019 12:28:59
1  720477  yandex  18.10.2019 12:53:05
2  720479      vk  18.10.2019 13:04:41
3  720483  google  18.10.2019 13:24:01

У DataFrame есть метод append, который позволяет добавить данные в конец таблицы:

In [5]:
Log_1.append(Log_2)

ID  SOURCE                 DATE
0   720251  yandex  17.10.2019 11:17:17
1   720250  google  17.10.2019 11:09:32
2   720255  yandex  17.10.2019 12:04:34
3   720257  google  17.10.2019 12:17:11
4   720258  yandex  17.10.2019 12:40:07
5   720281  yandex  17.10.2019 14:10:31
6   720290  yandex  17.10.2019 14:37:11
7   720297  yandex  17.10.2019 15:21:57
8   720301  yandex  17.10.2019 16:20:13
9   720303  yandex  17.10.2019 16:26:25
10  720253  google  17.10.2019 11:53:01
11  720256  yandex  17.10.2019 12:12:44
12  720262  yandex  17.10.2019 12:53:55
0   720474  yandex  18.10.2019 12:28:59
1   720477  yandex  18.10.2019 12:53:05
2   720479      vk  18.10.2019 13:04:41
3   720483  google  18.10.2019 13:24:01

*Примечание.* Заметьте, что индексы у наших записей остались от исходных фреймов. Если хотим пронумеровать записи заново, добавляем атрибут ignore_index=True

In [6]:
Log_1.append(Log_2, ignore_index=True)

ID  SOURCE                 DATE
0   720251  yandex  17.10.2019 11:17:17
1   720250  google  17.10.2019 11:09:32
2   720255  yandex  17.10.2019 12:04:34
3   720257  google  17.10.2019 12:17:11
4   720258  yandex  17.10.2019 12:40:07
5   720281  yandex  17.10.2019 14:10:31
6   720290  yandex  17.10.2019 14:37:11
7   720297  yandex  17.10.2019 15:21:57
8   720301  yandex  17.10.2019 16:20:13
9   720303  yandex  17.10.2019 16:26:25
10  720253  google  17.10.2019 11:53:01
11  720256  yandex  17.10.2019 12:12:44
12  720262  yandex  17.10.2019 12:53:55
13  720474  yandex  18.10.2019 12:28:59
14  720477  yandex  18.10.2019 12:53:05
15  720479      vk  18.10.2019 13:04:41
16  720483  google  18.10.2019 13:24:01

С помощью Append можно добавлять и отдельные строки (pandas.Series или похожие на словарь), но:

    1. Их надо передавать списком (даже если добавляем 1 строку, она должна быть в списке)
    2. Важно корректно задать ключи, иначе результат вас не обрадует.

In [7]:
Log_2.append([{'ID':1, 'SOURCE':'user', 'DATE':'now'}])

ID  SOURCE                 DATE
0  720474  yandex  18.10.2019 12:28:59
1  720477  yandex  18.10.2019 12:53:05
2  720479      vk  18.10.2019 13:04:41
3  720483  google  18.10.2019 13:24:01
0       1    user                  now

In [8]:
Log_2.append([(1, 'user', 'now')])

0     1    2                 DATE        ID  SOURCE
0  NaN   NaN  NaN  18.10.2019 12:28:59  720474.0  yandex
1  NaN   NaN  NaN  18.10.2019 12:53:05  720477.0  yandex
2  NaN   NaN  NaN  18.10.2019 13:04:41  720479.0      vk
3  NaN   NaN  NaN  18.10.2019 13:24:01  720483.0  google
0  1.0  user  now                  NaN       NaN     NaN

Другой способ объединить таблицы (произвольное их количество) - concat:

In [9]:
pd.concat((Log_1, Log_2))

ID  SOURCE                 DATE
0   720251  yandex  17.10.2019 11:17:17
1   720250  google  17.10.2019 11:09:32
2   720255  yandex  17.10.2019 12:04:34
3   720257  google  17.10.2019 12:17:11
4   720258  yandex  17.10.2019 12:40:07
5   720281  yandex  17.10.2019 14:10:31
6   720290  yandex  17.10.2019 14:37:11
7   720297  yandex  17.10.2019 15:21:57
8   720301  yandex  17.10.2019 16:20:13
9   720303  yandex  17.10.2019 16:26:25
10  720253  google  17.10.2019 11:53:01
11  720256  yandex  17.10.2019 12:12:44
12  720262  yandex  17.10.2019 12:53:55
0   720474  yandex  18.10.2019 12:28:59
1   720477  yandex  18.10.2019 12:53:05
2   720479      vk  18.10.2019 13:04:41
3   720483  google  18.10.2019 13:24:01

Заметим, что порядок таблиц, передаваемых для объединения важен и соответствует порядку строк в результате.

In [10]:
pd.concat((Log_2, Log_1))

ID  SOURCE                 DATE
0   720474  yandex  18.10.2019 12:28:59
1   720477  yandex  18.10.2019 12:53:05
2   720479      vk  18.10.2019 13:04:41
3   720483  google  18.10.2019 13:24:01
0   720251  yandex  17.10.2019 11:17:17
1   720250  google  17.10.2019 11:09:32
2   720255  yandex  17.10.2019 12:04:34
3   720257  google  17.10.2019 12:17:11
4   720258  yandex  17.10.2019 12:40:07
5   720281  yandex  17.10.2019 14:10:31
6   720290  yandex  17.10.2019 14:37:11
7   720297  yandex  17.10.2019 15:21:57
8   720301  yandex  17.10.2019 16:20:13
9   720303  yandex  17.10.2019 16:26:25
10  720253  google  17.10.2019 11:53:01
11  720256  yandex  17.10.2019 12:12:44
12  720262  yandex  17.10.2019 12:53:55

Так же заметим, что нумерация так же сохраняется от исходных таблиц. Избавиться от неё так же можно с помощью ignore_index=True.

In [11]:
pd.concat((Log_1, Log_2), ignore_index=True)

ID  SOURCE                 DATE
0   720251  yandex  17.10.2019 11:17:17
1   720250  google  17.10.2019 11:09:32
2   720255  yandex  17.10.2019 12:04:34
3   720257  google  17.10.2019 12:17:11
4   720258  yandex  17.10.2019 12:40:07
5   720281  yandex  17.10.2019 14:10:31
6   720290  yandex  17.10.2019 14:37:11
7   720297  yandex  17.10.2019 15:21:57
8   720301  yandex  17.10.2019 16:20:13
9   720303  yandex  17.10.2019 16:26:25
10  720253  google  17.10.2019 11:53:01
11  720256  yandex  17.10.2019 12:12:44
12  720262  yandex  17.10.2019 12:53:55
13  720474  yandex  18.10.2019 12:28:59
14  720477  yandex  18.10.2019 12:53:05
15  720479      vk  18.10.2019 13:04:41
16  720483  google  18.10.2019 13:24:01

Но что, если наши данные разделены не по датам, а по смыслу. Например, в одном файле информация об источнике заявки, а в другом результат её обработки (бизнес данные).

В таком случае нашей задачей будет объединить таблицы не построчно, а по столбцам.

In [12]:
Log_2_success = pd.read_csv('log_2_success.csv', sep=';')
Log_2_success

ID  SUCCESS
0  720474     True
1  720477    False
2  720479    False
3  720483     True

Для начала вызовем concat и посмотрим на результат.

Функция отработала, но предупреждает о вероятно некорректном результате:

In [13]:
pd.concat((Log_2, Log_2_success))

C:\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """Entry point for launching an IPython kernel.


DATE      ID  SOURCE SUCCESS
0  18.10.2019 12:28:59  720474  yandex     NaN
1  18.10.2019 12:53:05  720477  yandex     NaN
2  18.10.2019 13:04:41  720479      vk     NaN
3  18.10.2019 13:24:01  720483  google     NaN
0                  NaN  720474     NaN    True
1                  NaN  720477     NaN   False
2                  NaN  720479     NaN   False
3                  NaN  720483     NaN    True

Чтобы объединить таблицы по столбцам изменим ось (направление) объединения со строк (0) на столбцы (1) с помощью атрибута axis:

In [14]:
pd.concat((Log_2, Log_2_success), axis = 1)

ID  SOURCE                 DATE      ID  SUCCESS
0  720474  yandex  18.10.2019 12:28:59  720474     True
1  720477  yandex  18.10.2019 12:53:05  720477    False
2  720479      vk  18.10.2019 13:04:41  720479    False
3  720483  google  18.10.2019 13:24:01  720483     True

Т.к. столбец ID дублируется, давайте его удалим, просто сделав выборку нужных данных (напоминаем, объединять можно как фреймы так pandas.Series):

In [15]:
pd.concat((Log_2, Log_2_success['SUCCESS']), axis = 1)

ID  SOURCE                 DATE  SUCCESS
0  720474  yandex  18.10.2019 12:28:59     True
1  720477  yandex  18.10.2019 12:53:05    False
2  720479      vk  18.10.2019 13:04:41    False
3  720483  google  18.10.2019 13:24:01     True

До сих пор мы предполагали, что данные в обоих таблицах идут в одном и том же порядке. Но что, если они перепутаны?

**Следующий пример показывает некорректное объединение данных (нет синтаксической ошибки, но значения неверные, что легко проверить по несоответствию чисел в столбцах ID**

In [16]:
Log_2_success_unsorted = pd.read_csv('log_2_success_unsorted.csv', sep=';')
Log_2_success_unsorted

ID  SUCCESS
0  720477    False
1  720479    False
2  720474     True
3  720483     True

In [17]:
pd.concat((Log_2, Log_2_success_unsorted), axis = 1)

ID  SOURCE                 DATE      ID  SUCCESS
0  720474  yandex  18.10.2019 12:28:59  720477    False
1  720477  yandex  18.10.2019 12:53:05  720479    False
2  720479      vk  18.10.2019 13:04:41  720474     True
3  720483  google  18.10.2019 13:24:01  720483     True

Чтобы избежать этого, используем функцию merge аналогично предыдущей, но в параметр on передадим кортеж с именами ключей, в которых надо искать соответствие.

Т.е. функция будет брать строку из 1 таблицы, смотреть значение в ID, искать строку с **таким же** значением в ID во второй таблице и склеивать эти 2 строки. И так по всем строкам.

*Бонусом избавляемся от дублирования столбцов ID*

In [19]:
pd.merge(Log_2, Log_2_success_unsorted, on=('ID'))

ID  SOURCE                 DATE  SUCCESS
0  720474  yandex  18.10.2019 12:28:59     True
1  720477  yandex  18.10.2019 12:53:05    False
2  720479      vk  18.10.2019 13:04:41    False
3  720483  google  18.10.2019 13:24:01     True

А теперь давайте объединим 3 таблицы:
    
* Log_1 и Log_2 - логи за 2 дня
* Log_2_success_unsorted - результат успешности или неуспешности операций 2 дня
    
В этом примере у нас нет данных об успехе (или неуспехе) 1 дня, т.е. нет записей, соответсвующих этим строкам для объединения столбцов. 

In [20]:
pd.merge(Log_1.append(Log_2), Log_2_success_unsorted, on=('ID'))

ID  SOURCE                 DATE  SUCCESS
0  720474  yandex  18.10.2019 12:28:59     True
1  720477  yandex  18.10.2019 12:53:05    False
2  720479      vk  18.10.2019 13:04:41    False
3  720483  google  18.10.2019 13:24:01     True

Как видим, объединение произведено таким образом, что все строки, которым не нашлось парных записей для merge просто УДАЛЕНЫ.

Если мы хотим оставить их (планируем заполнить чем-то в будущем), то можно использовать метод объединения "left" (*Привет, SQL, Left Join!*) атрибутом how ='left':

In [21]:
pd.merge(Log_1.append(Log_2), Log_2_success_unsorted, on=('ID'), how ='left')

ID  SOURCE                 DATE SUCCESS
0   720251  yandex  17.10.2019 11:17:17     NaN
1   720250  google  17.10.2019 11:09:32     NaN
2   720255  yandex  17.10.2019 12:04:34     NaN
3   720257  google  17.10.2019 12:17:11     NaN
4   720258  yandex  17.10.2019 12:40:07     NaN
5   720281  yandex  17.10.2019 14:10:31     NaN
6   720290  yandex  17.10.2019 14:37:11     NaN
7   720297  yandex  17.10.2019 15:21:57     NaN
8   720301  yandex  17.10.2019 16:20:13     NaN
9   720303  yandex  17.10.2019 16:26:25     NaN
10  720253  google  17.10.2019 11:53:01     NaN
11  720256  yandex  17.10.2019 12:12:44     NaN
12  720262  yandex  17.10.2019 12:53:55     NaN
13  720474  yandex  18.10.2019 12:28:59    True
14  720477  yandex  18.10.2019 12:53:05   False
15  720479      vk  18.10.2019 13:04:41   False
16  720483  google  18.10.2019 13:24:01    True

Что если мы хотим объединить таблицы, находя соответствие не по колонкам, а по индексам?

In [22]:
Log_1_i = pd.read_csv('log_1.csv', sep=';', index_col='ID')
Log_2_i = pd.read_csv('log_2.csv', sep=';', index_col='ID')
Log_2_success_unsorted_i = pd.read_csv('log_2_success_unsorted.csv', sep=';', index_col='ID')

In [23]:
Log_2_i.head()

SOURCE                 DATE
ID                                 
720474  yandex  18.10.2019 12:28:59
720477  yandex  18.10.2019 12:53:05
720479      vk  18.10.2019 13:04:41
720483  google  18.10.2019 13:24:01

In [24]:
Log_2_success_unsorted_i.head()

SUCCESS
ID             
720477    False
720479    False
720474     True
720483     True

In [25]:
pd.merge(Log_1_i.append(Log_2_i), Log_2_success_unsorted_i, on=('ID'), how ='left')

SOURCE                 DATE SUCCESS
ID                                         
720251  yandex  17.10.2019 11:17:17     NaN
720250  google  17.10.2019 11:09:32     NaN
720255  yandex  17.10.2019 12:04:34     NaN
720257  google  17.10.2019 12:17:11     NaN
720258  yandex  17.10.2019 12:40:07     NaN
720281  yandex  17.10.2019 14:10:31     NaN
720290  yandex  17.10.2019 14:37:11     NaN
720297  yandex  17.10.2019 15:21:57     NaN
720301  yandex  17.10.2019 16:20:13     NaN
720303  yandex  17.10.2019 16:26:25     NaN
720253  google  17.10.2019 11:53:01     NaN
720256  yandex  17.10.2019 12:12:44     NaN
720262  yandex  17.10.2019 12:53:55     NaN
720474  yandex  18.10.2019 12:28:59    True
720477  yandex  18.10.2019 12:53:05   False
720479      vk  18.10.2019 13:04:41   False
720483  google  18.10.2019 13:24:01    True

К сожалению, это не сработает, если имена колонок/индексов разные (т.к. в атрибуте on задаётся имя общей колонки).

В таком случае правильнее использовать аргументы для объединения по индексам:
    
* left_index=True - для использования колонки индексов для объединения для левой таблицы (1-й аргумент)

* right_index=True - для использования колонки индексов для объединения для правой таблицы (2-й аргумент)

In [27]:
Log_2_success_unsorted_i.index.names = ['NEW_ID']
Log_2_success_unsorted_i

SUCCESS
NEW_ID         
720477    False
720479    False
720474     True
720483     True

In [28]:
pd.merge(Log_1_i.append(Log_2_i), Log_2_success_unsorted_i, how ='left', left_index=True, right_index=True)

SOURCE                 DATE SUCCESS
ID                                         
720251  yandex  17.10.2019 11:17:17     NaN
720250  google  17.10.2019 11:09:32     NaN
720255  yandex  17.10.2019 12:04:34     NaN
720257  google  17.10.2019 12:17:11     NaN
720258  yandex  17.10.2019 12:40:07     NaN
720281  yandex  17.10.2019 14:10:31     NaN
720290  yandex  17.10.2019 14:37:11     NaN
720297  yandex  17.10.2019 15:21:57     NaN
720301  yandex  17.10.2019 16:20:13     NaN
720303  yandex  17.10.2019 16:26:25     NaN
720253  google  17.10.2019 11:53:01     NaN
720256  yandex  17.10.2019 12:12:44     NaN
720262  yandex  17.10.2019 12:53:55     NaN
720474  yandex  18.10.2019 12:28:59    True
720477  yandex  18.10.2019 12:53:05   False
720479      vk  18.10.2019 13:04:41   False
720483  google  18.10.2019 13:24:01    True

Предположим, что у одной таблицы мы хотим использовать индекс, а у другой колонку для объединения. Тогда помимо указанных выше аргументов нам потребуется 1 (или оба) аргумента:
    
* left_on - имя колонки из левой таблицы (1-й аргумент) для объединения

* right_on - имя колонки из правой таблицы (2-й аргумент) для объединения

In [29]:
pd.merge(Log_1_i.append(Log_2_i), Log_2_success_unsorted_i, how ='left', left_on=('ID'), right_index=True)

SOURCE                 DATE SUCCESS
ID                                         
720251  yandex  17.10.2019 11:17:17     NaN
720250  google  17.10.2019 11:09:32     NaN
720255  yandex  17.10.2019 12:04:34     NaN
720257  google  17.10.2019 12:17:11     NaN
720258  yandex  17.10.2019 12:40:07     NaN
720281  yandex  17.10.2019 14:10:31     NaN
720290  yandex  17.10.2019 14:37:11     NaN
720297  yandex  17.10.2019 15:21:57     NaN
720301  yandex  17.10.2019 16:20:13     NaN
720303  yandex  17.10.2019 16:26:25     NaN
720253  google  17.10.2019 11:53:01     NaN
720256  yandex  17.10.2019 12:12:44     NaN
720262  yandex  17.10.2019 12:53:55     NaN
720474  yandex  18.10.2019 12:28:59    True
720477  yandex  18.10.2019 12:53:05   False
720479      vk  18.10.2019 13:04:41   False
720483  google  18.10.2019 13:24:01    True

### Связь один ко многим

Что если нет однозначного соответствия строк одной таблицы другой, например, одна из таблиц - это справочник?

Предположим, мы где-то разжились выгрузкой из 2 таблиц:

* orders.csv - информация по позициям в заказе (по 1 товару в строке, если в заказе несколько товаров, то в таблице будет несколько записей)

* Products.csv - выгрузка из справочника товаров (название, ID, цена и валюта)

Не все товары из Products.csv фигурируют в orders.csv.

Некоторые товары купили более одного раза.

In [30]:
orders = pd.read_csv('orders.csv', sep=';')
products = pd.read_csv('Products.csv', sep=';')

In [31]:
orders

Дата создания  Order ID  ID Покупателя  \
0   09.11.2019 21:55:51         9             10   
1   09.11.2019 15:05:57         8              9   
2   09.11.2019 15:05:57         8              9   
3   09.11.2019 12:50:07         7              8   
4   09.11.2019 12:00:00         6              1   
5   09.11.2019 12:00:00         6              1   
6   08.11.2019 08:36:22         5              5   
7   08.11.2019 08:36:22         4              9   
8   08.11.2019 08:36:22         3              8   
9   08.11.2019 08:36:22         3              8   
10  08.11.2019 08:36:21         2              7   
11  08.11.2019 08:36:21         2              7   
12  08.11.2019 08:36:21         2              7   
13  08.11.2019 08:36:21         2              7   
14  08.11.2019 08:36:20         1              5   
15  08.11.2019 08:36:20         1              5   
16  08.11.2019 08:36:20         1              5   
17  01.01.2001 00:00:00         0              1   

                             Статус Оплачен Отменен Отгружен  ID товара  \
0          Принят, ожидается оплата     Нет     Нет      Нет        103   
1          Принят, ожидается оплата     Нет     Нет      Нет         86   
2          Принят, ожидается оплата     Нет     Нет      Нет        104   
3          Принят, ожидается оплата     Нет     Нет      Нет        104   
4          Принят, ожидается оплата     Нет     Нет      Нет        104   
5          Принят, ожидается оплата     Нет     Нет      Нет        103   
6                           Отменён     Нет      Да      Нет        124   
7          Принят, ожидается оплата     Нет     Нет       Да         91   
8   Оплачен, формируется к отправке      Да     Нет      Нет        103   
9   Оплачен, формируется к отправке      Да     Нет      Нет        104   
10                         Выполнен      Да     Нет       Да         74   
11                         Выполнен      Да     Нет       Да         53   
12                         Выполнен      Да     Нет       Да        137   
13                         Выполнен      Да     Нет       Да         47   
14  Оплачен, формируется к отправке      Да     Нет      Нет         71   
15  Оплачен, формируется к отправке      Да     Нет      Нет         86   
16  Оплачен, формируется к отправке      Да     Нет      Нет         51   
17  Оплачен, формируется к отправке      Да     Нет      Нет        666   

    Количество  
0            5  
1          100  
2           10  
3            7  
4            5  
5            5  
6            1  
7            1  
8            3  
9            3  
10           1  
11           1  
12           1  
13           1  
14           1  
15           1  
16           1  
17           1

In [32]:
products

Product_ID                                   Name  Price CURRENCY
0           47                   Шатны Полосатый рейс   2999      RUR
1           51               Платье Аленький цветочек   4999      RUR
2           53                 Штаны Цветочная Поляна   4999      RUR
3           71                    Платье Ночная Жизнь   7999      RUR
4           74               Платье Ночная Жизнь XXXL   8999      RUR
5           86                     Носки Простые, муж     45      RUR
6           91                     Носки Честные, муж     50      RUR
7          103                  Носки Подарочные, муж    199      RUR
8          104                  Носки Подарочные, жен    249      RUR
9          124                   Носки беговые Camino    999      RUR
10         137              Гольфы детские Снегурочка     99      RUR
11         138               Гольфы детские Питер Пен     99      RUR
12         139                       Сумка для носков    500      RUR
13         140  Товар, которому не придумано название      1      RUR

Нам вновь понадобится merge.

In [33]:
pd.merge(orders, products, how ='left', left_on=('ID товара'), right_on=('Product_ID'))

Дата создания  Order ID  ID Покупателя  \
0   09.11.2019 21:55:51         9             10   
1   09.11.2019 15:05:57         8              9   
2   09.11.2019 15:05:57         8              9   
3   09.11.2019 12:50:07         7              8   
4   09.11.2019 12:00:00         6              1   
5   09.11.2019 12:00:00         6              1   
6   08.11.2019 08:36:22         5              5   
7   08.11.2019 08:36:22         4              9   
8   08.11.2019 08:36:22         3              8   
9   08.11.2019 08:36:22         3              8   
10  08.11.2019 08:36:21         2              7   
11  08.11.2019 08:36:21         2              7   
12  08.11.2019 08:36:21         2              7   
13  08.11.2019 08:36:21         2              7   
14  08.11.2019 08:36:20         1              5   
15  08.11.2019 08:36:20         1              5   
16  08.11.2019 08:36:20         1              5   
17  01.01.2001 00:00:00         0              1   

                             Статус Оплачен Отменен Отгружен  ID товара  \
0          Принят, ожидается оплата     Нет     Нет      Нет        103   
1          Принят, ожидается оплата     Нет     Нет      Нет         86   
2          Принят, ожидается оплата     Нет     Нет      Нет        104   
3          Принят, ожидается оплата     Нет     Нет      Нет        104   
4          Принят, ожидается оплата     Нет     Нет      Нет        104   
5          Принят, ожидается оплата     Нет     Нет      Нет        103   
6                           Отменён     Нет      Да      Нет        124   
7          Принят, ожидается оплата     Нет     Нет       Да         91   
8   Оплачен, формируется к отправке      Да     Нет      Нет        103   
9   Оплачен, формируется к отправке      Да     Нет      Нет        104   
10                         Выполнен      Да     Нет       Да         74   
11                         Выполнен      Да     Нет       Да         53   
12                         Выполнен      Да     Нет       Да        137   
13                         Выполнен      Да     Нет       Да         47   
14  Оплачен, формируется к отправке      Да     Нет      Нет         71   
15  Оплачен, формируется к отправке      Да     Нет      Нет         86   
16  Оплачен, формируется к отправке      Да     Нет      Нет         51   
17  Оплачен, формируется к отправке      Да     Нет      Нет        666   

    Количество  Product_ID                       Name   Price CURRENCY  
0            5       103.0      Носки Подарочные, муж   199.0      RUR  
1          100        86.0         Носки Простые, муж    45.0      RUR  
2           10       104.0      Носки Подарочные, жен   249.0      RUR  
3            7       104.0      Носки Подарочные, жен   249.0      RUR  
4            5       104.0      Носки Подарочные, жен   249.0      RUR  
5            5       103.0      Носки Подарочные, муж   199.0      RUR  
6            1       124.0       Носки беговые Camino   999.0      RUR  
7            1        91.0         Носки Честные, муж    50.0      RUR  
8            3       103.0      Носки Подарочные, муж   199.0      RUR  
9            3       104.0      Носки Подарочные, жен   249.0      RUR  
10           1        74.0   Платье Ночная Жизнь XXXL  8999.0      RUR  
11           1        53.0     Штаны Цветочная Поляна  4999.0      RUR  
12           1       137.0  Гольфы детские Снегурочка    99.0      RUR  
13           1        47.0       Шатны Полосатый рейс  2999.0      RUR  
14           1        71.0        Платье Ночная Жизнь  7999.0      RUR  
15           1        86.0         Носки Простые, муж    45.0      RUR  
16           1        51.0   Платье Аленький цветочек  4999.0      RUR  
17           1         NaN                        NaN     NaN      NaN

Можно заметить на 17й строке заказ ID=0 со странной товарной позицией 666, которой просто нет в справочнике. Это явно тестовая запись, которой не нашлось соответствия.

Удалить её можно вручную, по NaN в одной из колонок, либо с помощью объединения how='inner' (тогда строки из левой таблицы, которым не найдено соответствие в правой, так же будут исключены из результата).

In [34]:
pd.merge(orders, products, how='inner', left_on=('ID товара'), right_on=('Product_ID'))

Дата создания  Order ID  ID Покупателя  \
0   09.11.2019 21:55:51         9             10   
1   09.11.2019 12:00:00         6              1   
2   08.11.2019 08:36:22         3              8   
3   09.11.2019 15:05:57         8              9   
4   08.11.2019 08:36:20         1              5   
5   09.11.2019 15:05:57         8              9   
6   09.11.2019 12:50:07         7              8   
7   09.11.2019 12:00:00         6              1   
8   08.11.2019 08:36:22         3              8   
9   08.11.2019 08:36:22         5              5   
10  08.11.2019 08:36:22         4              9   
11  08.11.2019 08:36:21         2              7   
12  08.11.2019 08:36:21         2              7   
13  08.11.2019 08:36:21         2              7   
14  08.11.2019 08:36:21         2              7   
15  08.11.2019 08:36:20         1              5   
16  08.11.2019 08:36:20         1              5   

                             Статус Оплачен Отменен Отгружен  ID товара  \
0          Принят, ожидается оплата     Нет     Нет      Нет        103   
1          Принят, ожидается оплата     Нет     Нет      Нет        103   
2   Оплачен, формируется к отправке      Да     Нет      Нет        103   
3          Принят, ожидается оплата     Нет     Нет      Нет         86   
4   Оплачен, формируется к отправке      Да     Нет      Нет         86   
5          Принят, ожидается оплата     Нет     Нет      Нет        104   
6          Принят, ожидается оплата     Нет     Нет      Нет        104   
7          Принят, ожидается оплата     Нет     Нет      Нет        104   
8   Оплачен, формируется к отправке      Да     Нет      Нет        104   
9                           Отменён     Нет      Да      Нет        124   
10         Принят, ожидается оплата     Нет     Нет       Да         91   
11                         Выполнен      Да     Нет       Да         74   
12                         Выполнен      Да     Нет       Да         53   
13                         Выполнен      Да     Нет       Да        137   
14                         Выполнен      Да     Нет       Да         47   
15  Оплачен, формируется к отправке      Да     Нет      Нет         71   
16  Оплачен, формируется к отправке      Да     Нет      Нет         51   

    Количество  Product_ID                       Name  Price CURRENCY  
0            5         103      Носки Подарочные, муж    199      RUR  
1            5         103      Носки Подарочные, муж    199      RUR  
2            3         103      Носки Подарочные, муж    199      RUR  
3          100          86         Носки Простые, муж     45      RUR  
4            1          86         Носки Простые, муж     45      RUR  
5           10         104      Носки Подарочные, жен    249      RUR  
6            7         104      Носки Подарочные, жен    249      RUR  
7            5         104      Носки Подарочные, жен    249      RUR  
8            3         104      Носки Подарочные, жен    249      RUR  
9            1         124       Носки беговые Camino    999      RUR  
10           1          91         Носки Честные, муж     50      RUR  
11           1          74   Платье Ночная Жизнь XXXL   8999      RUR  
12           1          53     Штаны Цветочная Поляна   4999      RUR  
13           1         137  Гольфы детские Снегурочка     99      RUR  
14           1          47       Шатны Полосатый рейс   2999      RUR  
15           1          71        Платье Ночная Жизнь   7999      RUR  
16           1          51   Платье Аленький цветочек   4999      RUR